libminifi/include/core/logging/Logger.h

```diff
@@ -53,28 +54,32 @@ class LoggerControl {
   std::atomic<bool> is_enabled_;
 };
 
-inline char const* conditional_conversion(std::string const& str) {
-  return str.c_str();
-}
+template<typename Arg>
+using has_const_c_str_method = decltype(std::declval<const Arg&>().c_str());
 
-template<size_t N>
-inline char const* conditional_conversion(const utils::SmallString<N>& arr) {
-  return arr.c_str();
+template<typename Arg>
+inline decltype(auto) conditional_stringify(Arg&& arg) {
+  if constexpr (utils::meta::is_detected_v<has_const_c_str_method, Arg> || std::is_scalar_v<std::decay_t<Arg>>) {
+    return std::forward<Arg>(arg);
+  } else if constexpr (std::is_invocable_v<Arg>) {
+    return std::forward<Arg>(arg)();
```

Comment:
since we call `std::is_invocable_v` with zero arguments, it is exactly the zero argument callable FunctionObjects it is checking for

Reply:
It's checking for zero argument `Callable`, not `FunctionObject`.
```suggestion
  } else if constexpr (std::is_invocable_v<Arg>) {
    return std::invoke(std::forward<Arg>(arg));
```

Reply:
> otherwise, `INVOKE(f, t1, t2, ..., tN)` is equivalent to `f(t1, t2, ..., tN)` (that is, f is a [FunctionObject](https://en.cppreference.com/w/cpp/named_req/FunctionObject))

Reply:
could you give an example zero argument `Callable` that is not a `FunctionObject`?

Reply:
`&C2Payload::validate`